src/dotnet/BuildServer/RazorAssemblyResolver.cs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Build.Execution;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.BuildServer
+{
+    internal class RazorAssemblyResolver : IRazorAssemblyResolver
+    {
+        private readonly IDirectory _directory;
+
+        public RazorAssemblyResolver(IDirectory directory = null)
+        {
+            _directory = directory ?? FileSystemWrapper.Default.Directory;
+        }
+
+        public IEnumerable<FilePath> EnumerateRazorToolAssemblies()
+        {
+            HashSet<string> seen = new HashSet<string>();
+
+            var globalProperties = new Dictionary<string, string>
+            {
+                // This property disables default item globbing to improve performance
+                // This should be safe because we are not evaluating items, only properties
+                { Constants.EnableDefaultItems, "false" }
+            };
+
+            foreach (var projectFile in _directory.EnumerateFiles(Directory.GetCurrentDirectory(), "*.*proj"))
```

Comment:
Hmm. I wasn't expecting CWD nor the state of projects therein to impact this. Can we not use a well-known path? Is this tool in the nuget cache or something?

Reply:
It is in the NuGet cache.

Reply:
e.g. `.nuget/packages/microsoft.aspnetcore.razor.design/2.1.0-preview2-30475/tools/rzc.dll`

Reply:
I don't know the right fix here, but I'm thinking this doesn't actually work with the general CI workflow:

```
dotnet build [sln]
dotnet buildserver shutdown
```

The use of Razor would be buried in the build of the solution, not in a *.*proj in current directory.

Reply:
@rynowak thoughts?  We could have shutdown take a path to a project file.

Reply:
Hmm, there's a related issue here for vbcscompiler if user is using a custom version of Microsoft.NETCore.Compilers, which I hear will be supported officially soon.

I think in all of the discussion leading up to this there was an assumption that the command is global to an SDK version, not a function of what's in your project files.

It almost feels like we need a protocol to traverse sln and find all tools to shutdown, etc. But that is going to add a bunch of evaluation time just to shutdown, and is a heck of a lot more complicated than what we were imagining when the issue was filed.

cc @agocke @jaredpar

Reply:
I might be late to this discussion, but what if we used something a little more cooperative. The idea of finding all servers running and telling them each to shut down seems a little fraught.

This could be something like all listening to a specific pipe in servers, or all watching for a specific sentinel file to be created.

> It almost feels like we need a protocol to traverse sln and find all tools to shutdown, etc. But that is going to add a bunch of evaluation time just to shutdown, and is a heck of a lot more complicated than what we were imagining when the issue was filed.

Agree and agree. sln/project are concepts that users already understand and are a good hook for more msbuild magic.

Reply:
I think this approach is acceptable for preview2, but I would like to discuss (offline) with everyone how to better discover build-related servers so we can shut them down in a more unified manner.

Reply:
I'm OK with that.